### PR TITLE
feat: add C `ndarray` API and refactor `blas/ext/base/dnansumkbn`

### PR DESCRIPTION
**Pull request by headlessNode**
Resolves none.

## Description

> What is the purpose of this pull request?

This pull request:

-   Adds C `ndarray` API and refactor/cleanup `blas/ext/base/dnansumkbn`

## Related Issues

> Does this pull request have any related issues?

No. 

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

### CODE REVIEW

File: lib/node_modules/@stdlib/blas/ext/base/dnansumkbn/lib/ndarray.js [hunk 1]
```diff
@@ -51,26 +51,27 @@ var abs = require( '@stdlib/math/base/special/abs' );
...
 	var t;
 	var c;
 	var i;
 
+	ix = offsetX;
+	sum = 0.0;
+	c = 0.0;
 	if ( N <= 0 ) {
 		return 0.0;
 	}
```
Comment (headlessNode) [inline lib/node_modules/@stdlib/blas/ext/base/dnansumkbn/lib/ndarray.js]:
```suggestion
	sum = 0.0;
	c = 0.0;
	if ( N <= 0 ) {
		return 0.0;
	}
	ix = offsetX;
```

File: lib/node_modules/@stdlib/blas/ext/base/dnansumkbn/src/main.c [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+	double v;
+	double t;
+	double c;
+
+	sum = 0.0;
+	c = 0.0;
+	ix = offsetX;
+	if ( N <= 0 ) {
+		return sum;
+	}
```
Comment (headlessNode) [inline lib/node_modules/@stdlib/blas/ext/base/dnansumkbn/src/main.c]:
```suggestion
	sum = 0.0;
	c = 0.0;
	if ( N <= 0 ) {
		return sum;
	}
	ix = offsetX;
```

File: lib/node_modules/@stdlib/blas/ext/base/dnansumkbn/lib/ndarray.js [hunk 3]
```diff
@@ -51,26 +51,27 @@ var abs = require( '@stdlib/math/base/special/abs' );
...
+function dnansumkbn( N, x, strideX, offsetX ) {
 	var sum;
 	var ix;
 	var v;
 	var t;
 	var c;
 	var i;
 
+	sum = 0.0;
+	c = 0.0;
```
Comment (headlessNode) [inline lib/node_modules/@stdlib/blas/ext/base/dnansumkbn/lib/ndarray.js]:
```suggestion
	sum = 0.0;
```

File: lib/node_modules/@stdlib/blas/ext/base/dnansumkbn/lib/ndarray.js [hunk 4]
```diff
@@ -51,26 +51,27 @@ var abs = require( '@stdlib/math/base/special/abs' );
...
+			return sum;
 		}
-		return x[ offset ];
+		sum = x[ ix ] * N;
+		return sum;
 	}
-	ix = offset;
-	sum = 0.0;
-	c = 0.0;
 	for ( i = 0; i < N; i++ ) {
```
Comment (headlessNode) [inline lib/node_modules/@stdlib/blas/ext/base/dnansumkbn/lib/ndarray.js:74]:
```suggestion
	c = 0.0;
	for ( i = 0; i < N; i++ ) {
```

File: lib/node_modules/@stdlib/blas/ext/base/dnansumkbn/src/main.c [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+	ix = offsetX;
+	if ( strideX == 0 ) {
+		if ( stdlib_base_is_nan( X[ ix ] ) ) {
+			return sum;
+		}
+		sum = X[ ix ] * N;
+		return sum;
+	}
+	for ( i = 0; i < N; i++ ) {
```
Comment (headlessNode) [inline lib/node_modules/@stdlib/blas/ext/base/dnansumkbn/src/main.c]:
```suggestion
	sum = 0.0;
	if ( N <= 0 ) {
		return sum;
	}
	ix = offsetX;
	if ( strideX == 0 ) {
		if ( stdlib_base_is_nan( X[ ix ] ) ) {
			return sum;
		}
		sum = X[ ix ] * N;
		return sum;
	}
	c = 0.0;
	for ( i = 0; i < N; i++ ) {
```

**Comment by kgryte** [comment]
@headlessNode The downstream test failure is expected in this PR?

**Comment by headlessNode** [comment]
@kgryte Yes, the `dnansum` uses `dnansumkbn` for computing the sum. Should be fixed when `dnansum` is refactored.

File: lib/node_modules/@stdlib/blas/ext/base/dnansumkbn/lib/index.js [hunk 6]
```diff
@@ -38,7 +38,7 @@
 *
 * var x = new Float64Array( [ 2.0, 1.0, 2.0, -2.0, -2.0, 2.0, 3.0, 4.0, NaN, NaN ] );
 *
-* var v = dnansumkbn.ndarray( 5, x, 2, 1 );
+* var v = dnansumkbn.ndarray( x.length, x, 2, 1 );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/ext/base/dnansumkbn/lib/index.js]:
```suggestion
* var v = dnansumkbn.ndarray( 5, x, 2, 1 );
```

This needs to remain `5`, as the stride equals `2`.

File: lib/node_modules/@stdlib/blas/ext/base/dnansumkbn/lib/ndarray.js [hunk 7]
```diff
@@ -51,25 +51,25 @@ var abs = require( '@stdlib/math/base/special/abs' );
...
 	var sum;
 	var ix;
 	var v;
 	var t;
 	var c;
 	var i;
 
+	sum = 0.0;
 	if ( N <= 0 ) {
 		return 0.0;
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/ext/base/dnansumkbn/lib/ndarray.js]:
```suggestion
		return sum;
```

File: lib/node_modules/@stdlib/blas/ext/base/dnansumkbn/src/main.c [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+* limitations under the License.
+*/
+
+#include "stdlib/blas/ext/base/dnansumkbn.h"
+#include "stdlib/strided/base/stride2offset.h"
+#include "stdlib/math/base/assert/is_nan.h"
+#include "stdlib/math/base/special/abs.h"
+#include "stdlib/blas/base/shared.h"
+
+
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/ext/base/dnansumkbn/src/main.c]:
```suggestion
```

**Review by kgryte — APPROVED** [review]
LGTM. Thanks, @headlessNode!

**Comment by Planeshifter** [comment]
Will merge shortly. 

Failing CI checks are due to a) `cppcheck` errors and b) `dnansum` unit test being in need of updating due to the equivalent PR of this one for `dnansumors`, which also changed the behavior for a stride of zero.

**Pull request merged**